rfcs/shared/build-system/08-type-checking-perf-improvements.md

```diff
@@ -0,0 +1,287 @@
+
+### 1. generate problematic(circular) dependencies types(`.d.ts`) first manually before `type-check` runs
+
+We can imperatively traverse graph and generate `d.ts` for packages that cause circular dependencies as can bee seen in this [PR](https://github.com/microsoft/fluentui/pull/28002/files#diff-38d7f0411164173eb390d7191ad068bfe70935082bc37198b6b35b00408b2b42).
+
+This dts generation can be done in 2 ways:
+
+**1. generate tasks first, then run builds**
+
+> EXAMPLE Logic: https://github.com/microsoft/fluentui/pull/28002/files#diff-7915b9b726a397ae7ba6af7b9703633d21c031ebf21682f3ee7e6a4ec52837a5
+>
+> EXAMPLE PR: https://github.com/microsoft/fluentui/pull/28002
+
+```yml
+- script: node ./scripts/executors/type-check-ci-hack.js --base $(targetBranch)
+    displayName: type-check perf preparation
+- script: yarn buildci $(sinceArg)
+```
+
+**Pros:**
+
+- none
+
+**Cons:**
+
+- bad DX for local machine
+
+**2. Leverage task graph orchestration and define dependency on every v9 `type-check`**
+
+> EXAMPLE Logic: https://github.com/microsoft/fluentui/pull/30577/files#diff-624201bdc18cf7b31fdca43ca08549cbd4db08cf5fa3fc9dfb51b0141669b1b8
+>
+> EXAMPLE PR: https://github.com/microsoft/fluentui/pull/30577
+
+```jsonc
+// @filename <package-root>/project.json
+{
+  "targets": {
+    "type-check": {
+      "dependsOn": [
+        /* global nx task/script that will be run automatically - its cached so it will do work only once  */
+        "generate-circular-deps-dts"
+      ]
+    }
+  }
+}
+```
+
+With this the flow would be
+
+```mermaid
+flowchart LR
+ type-check --> generate-circular-deps-dts --> build[^build]
+```
+
+**Pros:**
+
+- declarative approach
+- same behaviour on CI and local machine (NOTE: we run into race-conditions within MVP)
+
+**Cons:**
+
+- all problematic packages needs to override `targets#type-check` with `dependsOn` declaration
+- problems with current task orchestrator caching
+- huge peak of CPU usage
+
+#### Solution overall Pros and Cons
+
+**Pros:**
+
+- no package scaffold changes needed
+
+**Cons:**
+
+- brittle solution for future library requirements
+- manual processing and resolution of dependency graph + maintenance
+- adding more technical depth and magic on build layer
+- "hack" that is not solving circular dependencies
+
+### 2. get rid of circular dependencies
+
+We have 2 options on how to get rid of circular dependencies
+
+- stop importing from react-components suite within stories
+- move/encapsulate problematic parts of code to separate packages.
+
+#### 2.1 stop importing from react-components suite within stories
+
+```diff
+-import {Button} from '@fluentui/react-components'
++import {Button} from '@fluentui/react-button'
+```
+
+##### Solution overall Pros and Cons
+
+**Pros:**
+
+- no magic resolution of dependency tree behind the scenes
+
+**Cons:**
+
+- might run into circulars in the future so this approach won't scale
+- AST transforms needed for storybook rendering
```

Comment:
We already doing that for `@fluentui/react-*` packages in `@fluentui/react-storybook-addon-export-to-sandbox` via `@fluentui/babel-preset-storybook-full-source`:

### input

<img width="829" alt="image" src="https://github.com/microsoft/fluentui/assets/14183168/0efa0160-cc76-41ec-9c3b-a5a46489b7d0">

### output

<img width="520" alt="image" src="https://github.com/microsoft/fluentui/assets/14183168/f03fa3d9-6ebc-4d45-b7b3-091c6f1d2d3c">